subprojects/ide/src/main/java/org/gradle/plugins/ide/internal/resolver/IdeDependencySet.java

```diff
@@ -216,12 +218,19 @@ private void visitArtifacts(IdeDependencyVisitor visitor) {
                     Set<ResolvedArtifactResult> javaDoc = auxiliaryArtifacts.get(componentIdentifier, JavadocArtifact.class);
                     javaDoc = javaDoc != null ? javaDoc : Collections.<ResolvedArtifactResult>emptySet();
                     visitor.visitModuleDependency(artifact, sources, javaDoc, isTestConfiguration(configurations.get(artifactIdentifier)));
+                } else if (isGradleApiDependency(artifact)) {
+                    visitor.visitGradleApiDependency(artifact, gradleApiSourcesResolver.resolveGradleApiSources(artifact.getFile()), isTestConfiguration(configurations.get(artifactIdentifier)));
                 } else {
                     visitor.visitFileDependency(artifact, isTestConfiguration(configurations.get(artifactIdentifier)));
                 }
             }
         }
 
+        private boolean isGradleApiDependency(ResolvedArtifactResult artifact) {
+            String artifactFileName = artifact.getFile().getName();
+            return artifactFileName.startsWith("gradle-api") || artifactFileName.startsWith("gradle-test-kit");
```

Comment:
I think Donat had the same comment around using the `artifactFileName` to find the API Jar may be fragile. However, I think it's fair given there isn't any way to work around this at the moment.